src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,369 @@
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <chrono>
+
+#include "absl/container/btree_map.h"
+#include "absl/strings/str_format.h"
+#include "ray/common/grpc_util.h"
+#include "ray/rpc/client_call.h"
+#include "ray/rpc/grpc_client.h"
+
+namespace ray {
+namespace rpc {
+
+#define INVOKE_RETRYABLE_RPC_CALL(retryable_rpc_client,                       \
+                                  SERVICE,                                    \
+                                  METHOD,                                     \
+                                  request,                                    \
+                                  callback,                                   \
+                                  rpc_client,                                 \
+                                  method_timeout_ms)                          \
+  (retryable_rpc_client->CallMethod<SERVICE, METHOD##Request, METHOD##Reply>( \
+      &SERVICE::Stub::PrepareAsync##METHOD,                                   \
+      rpc_client,                                                             \
+      #SERVICE ".grpc_client." #METHOD,                                       \
+      request,                                                                \
+      callback,                                                               \
+      method_timeout_ms))
+
+// Define a void retryable RPC client method.
+#define VOID_RETRYABLE_RPC_CLIENT_METHOD(                                        \
+    retryable_rpc_client, SERVICE, METHOD, rpc_client, method_timeout_ms, SPECS) \
+  void METHOD(const METHOD##Request &request,                                    \
+              const ClientCallback<METHOD##Reply> &callback) SPECS {             \
+    INVOKE_RETRYABLE_RPC_CALL(retryable_rpc_client,                              \
+                              SERVICE,                                           \
+                              METHOD,                                            \
+                              request,                                           \
+                              callback,                                          \
+                              rpc_client,                                        \
+                              method_timeout_ms);                                \
+  }
+
+/// \class Executor
+/// Executor saves operation and support retries.
+class Executor {
+ public:
+  Executor(std::function<void(const ray::Status &)> abort_callback)
+      : abort_callback_(std::move(abort_callback)) {}
+
+  /// This function is used to execute the given operation.
```

Comment:
Documentation is not precise, better to mention we also store the function for retry?